lib/Sema/TypeCheckConcurrency.cpp

```diff
@@ -637,6 +637,42 @@ static bool isAsyncCall(const ApplyExpr *call) {
   return funcType->isAsync();
 }
 
+/// Extract the correct expression for diagnosing actor state being passed inout
+///
+/// This unwraps most expression types, working toward an offending DeclRefExpr
+/// or to the LookupExpr before the DeclRef.
+///
+/// The return will either be a LookupExpr or a DeclRefExpr.
+/// If part of the sub-expression isn't interesting, returns nullptr
+static const Expr *getInoutTopExpr(const InOutExpr *arg) {
```

Comment:
I wonder if, instead of trying to manually enumerate all of the possible expressions that could appear, you could try using some existing infrastructure that can handle some or all of those cases for you in this search for a `DeclRef`.

For example, we have `Expr::getReferencedDecl` which appears to do something similar to your `getInoutTopExpr` function. A key thing to note is that `getReferencedDecl` would return the referenced member, and not the base, of a `LookupExpr` (see the cases for `MemberRef`, as an example). I don't think that difference would limit you from using it in some capacity, though.

There's also `Expr::forEachChildExpr` which one could use to do a generic walk over the whole expression tree though perhaps that's a bit much?

Reply:
Okay, I've switched it to using `expr::forEachChildExpr`. Still a little nervous that it might accept invalid code.
Not a huge fan of the double-lambdas, but it should do the right things.